cmd/genji/insert.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/genjidb/genji"
+	"github.com/genjidb/genji/document"
+	"github.com/genjidb/genji/engine"
+	"github.com/genjidb/genji/engine/badgerengine"
+	"github.com/genjidb/genji/engine/boltengine"
+)
+
+func skipSpaces(r *bufio.Reader) (byte, error) {
+	var c byte
+	var err error
+
+	for {
+		c, err = r.ReadByte()
+		if err != nil {
+			return c, err
+		}
+
+		if c != '\n' && c != '\r' && c != ' ' && c != '\t' {
+			break
+		}
+	}
+
+	return c, nil
+}
+
+func executeInsertCommand(db *genji.DB, table string, r io.Reader) error {
+	q := fmt.Sprintf("INSERT INTO %s VALUES ?", table)
+	rd := bufio.NewReader(r)
+	var c byte
+	var err error
+
+	// Ignore spaces.
+	c, err = skipSpaces(rd)
+	if err != nil {
+		return err
+	}
+	switch c {
+	case '{': // Json objects
+		if err := rd.UnreadByte(); err != nil {
+			return err
+		}
+
+		dec := json.NewDecoder(rd)
+		// NewFromDecoder is a wrapper to call parseJsonDocument.
+		d, err := document.NewFromDecoder(dec)
+		if err != nil {
+			return err
+		}
+
+		if err := db.Exec(q, &d); err != nil {
+			return err
+		}
+	case '[': // Array of json objects
+		if err := rd.UnreadByte(); err != nil {
+			return err
+		}
+
+		dec := json.NewDecoder(rd)
+		// Unmarshall json returns []interface{} for JSON arrays.
+		var m []interface{}
```

Comment:
This will load every json document in memory. To give you an example, imagine inserting a json file of 100 GB and passing it to this program like this:

```bash
cat veryBigFile.json | genji insert ...
```

The current code will read all the file content into `m`, and unless we have 100 gb ram computers, the program will crash.
Instead, we want to read chunk by chunk, to keep the memory usage very low, and the json decoder does exactly that.
You have exactly what you need in this example, just copy paste it and adapt it to this function: https://golang.org/pkg/encoding/json/#Decoder